src/UbiHttp.cpp

```diff
@@ -140,102 +145,163 @@ bool UbiHTTP::sendData(const char* device_label, const char* device_name, char*
+    _client_https_ubi.stop();
+    return ERROR_VALUE;
+  }
 
-    _client_https_ubi.print(F("GET /api/v1.6/devices/"));
-    _client_https_ubi.print(device_label);
-    _client_https_ubi.print("/");
-    _client_https_ubi.print(variable_label);
-    _client_https_ubi.print("/lv");
-    _client_https_ubi.print(" HTTP/1.1\r\n");
-    _client_https_ubi.print("Host: ");
-    _client_https_ubi.print(_host);
-    _client_https_ubi.print("\r\n");
-    _client_https_ubi.print("User-Agent: ");
-    _client_https_ubi.print(_user_agent);
-    _client_https_ubi.print("\r\n");
-    _client_https_ubi.print("X-Auth-Token: ");
-    _client_https_ubi.print(_token);
-    _client_https_ubi.print("\r\n");
-    _client_https_ubi.print("Content-Type: application/json\r\n\r\n");
+  uint16_t pathLength = _pathLength(device_label, variable_label);
+  char *path = (char *)malloc(sizeof(char) * pathLength + 1);
+  sprintf(path, "/api/v1.6/devices/%s/%s/lv", device_label, variable_label);
+
+  if (_debug) {
+    Serial.print(F("Requesting to URL: "));
+    Serial.println(path);
+  }
+
+  uint16_t requestLineLength = _requestLineLength(path);
+  char *message = (char *)malloc(sizeof(char) * requestLineLength + 1);
+  sprintf(message,
+          "GET %s HTTP/1.1\r\nHost: %s\r\nX-Auth-Token: "
+          "%s\r\nUser-Agent: %s\r\nContent-Type: "
+          "application/json\r\nConnection: close\r\n\r\n",
+          path, _host, _token, _user_agent);
+
+  if (_debug) {
+    Serial.println(F("Request sent"));
+    Serial.println(message);
+  }
+
+  _client_https_ubi.print(message);
 
-    /* Waits for the host's answer */
-    if (!waitServerAnswer()) {
-      _client_https_ubi.stop();
-      return ERROR_VALUE;
+  while (_client_https_ubi.connected()) {
+    String line = _client_https_ubi.readStringUntil('\n');
+    if (line == "\r") {
+      if (_debug) {
+        Serial.println(F("Headers received"));
+      }
+      break;
     }
+  }
 
-    /* Reads the response from the server */
-    char* response = (char*)malloc(sizeof(char) * MAX_BUFFER_SIZE);
-    readServerAnswer(response);
+  free(message);
+  free(path);
 
-    /* Parses the answer */
-    _client_https_ubi.stop();
-    float value = parseHttpAnswer("LV", response);
-    free(response);
-    return value;
+  double value = _parseServerAnswer();
+
+  _client_https_ubi.flush();
+  _client_https_ubi.stop();
+  return value;
+}
+
+double UbiHTTP::_parseServerAnswer() {
+
+  /**
+   * @param _charResponseLength 3 is the maximun amount of digits for the length
+   * to have
+   */
+  char *_charLength = (char *)malloc(sizeof(char) * 3);
+
+  readServerAnswer(_charLength);
+
+  /**
+   * The server respond the length of the value in HEX so it has to be converted
+   * to DEC
+   * */
+  uint8_t length = UbiUtils::hexadecimalToDecimal(_charLength);
+
+  if (_debug) {
+    Serial.printf_P("Length: %i\n", length);
   }
 
+  char *_charValue = (char *)malloc(sizeof(char) * length + 1);
+
+  readServerAnswer(_charValue);
```

Comment:
why do you need to read twice the server answer? this should be done just once

Reply:
The readServerAnswer just parse the first line of the server response and return it ready to convert it to something readable, the code is quite explicit on what is trying to do.